proposals/0000-surface-async-stream-errors-to-http.md

```diff
@@ -0,0 +1,231 @@
+- https://nodejs.org/api/http.html#http_event_clienterror
+- `http2session.destroy([error][, code])`
+- `http2session.goaway([code[, lastStreamID[, opaqueData]]])`
+- https://nodejs.org/api/http2.html#http2_destruction
+- `http2stream.close(code[, callback])`
+- `response.addTrailers(headers)`
+
+### HTTP ecosystem considerations
+
+It should be possible for developers to selectively turn this error-signaling behavior off, in the case of misbehaving proxies or known-bad clients.
+
+Research needed into how common reverse proxies or load balancers, such as nginx, handle error-signaled HTTP streams. Real-world examples of how that can happen: https://rhodesmill.org/brandon/2013/chunked-wsgi/ (including the comment at the very bottom)
+
+https://rijulaggarwal.wordpress.com/2018/01/10/atmosphere-long-polling-on-nginx-chunked-encoding-error/
+
+In the common case of the backend speaking HTTP/1.1 to a reverse proxy/front-end terminator/CDN/etc. that translates to HTTP/2 for browsers, what can be done?
+
+## Migration strategy
+
+> Is this proposal backwards incompatible? Does this proposal replace an existing feature/pattern? If so, can we safely and automatically migrate existing apps/components to this proposal? How?
+
+> If applicable, provide sample error messages, deprecation warnings, or migration guidance.
+
+## Drawbacks
+
+So far, Marko is totally agnostic as to what kind of stream it’s outputting to. You could use it to stream to a file, or something really unexpected like an FTP stream transmission or a chat protocol.
+
+It’s possible the server runtime might expose a different method to avoid breaking such cases, like `template.streamToHttp()`.
+
+However, there’s already a lot about Marko’s server rendering that only makes sense delivered over HTTP. For example, if you were rendering to a file to be served later, you’d want to buffer the component init scripts and you’d never want `client-reorder` enabled.
+
+## Alternatives
+
+The impact of not doing this would be no change; we would continue exposing Marko authors to the risks mentioned in the Motivation section.
+
+In the past, browsers other than Internet Explorer accepted `multipart/x-mixed-replace` responses to be loaded in `target="_top"`, and that could potentially be used to swap to a a new copy of the response-in-progress but with error-indicating headers, or potentially even a replacement error page. However, modern browsers have quietly dropped support for `multipart/x-mixed-replace` in the top browsing context.
+
+It _might_ be possible to write `<meta http-equiv="refresh" content="0;url=${errorPageLocation}">`. It’s not _supposed_ to be outside the `<head>`, but I still like it better than how Rails creates a mid-stream error redirect with an inline `<script>`. ([some context on how search engines could better understand this is a temporary redirect, not a permanent one](https://twitter.com/JohnMu/status/969486943351394304))
```

Comment:
FWIW this is what I've seen in several applications. Curious if this is one of those "kinda not supposed to work but does" things.

Reply:
It definitely works for browsers and probably search engines, but caches still probably need the explicit indicator that the HTTP response was incomplete.

Reply:
One thing some of the prior art has handled partially is what happens if the stream error occurs while the HTML has already opened a tag, attribute value, etc. Rails prepends `">` and I’ve also seen `</script>`

GitHub explored this problem even more thoroughly for [Dangling Markup Busting in their Post-CSP journey](https://github.blog/2017-01-19-githubs-post-csp-journey/#dangling-markup-busting)

Reply:
Spitballing here:

```js
out.streamErrorRedirect = function (url) {
  out.shouldMarkAsIncomplete = true;
  out.write(`
    <!-- '" --><!-- </style></script></textarea></select></template> -->
    <meta http-equiv="refresh" content="0;url=${url}">
    <script>window.location = "${url}"</script>
  `);
};
```

- Should there be a default error page URL taken from `out.global.error500Url`, like `out.global.cspNonce`?

- How much should Marko protect against dangling markup? Rails and other existing solutions don’t seem to care much.

- Writing further error content should still happen with whatever is inside `<@catch>`, in case both the inline `<script>` and `meta[http-equiv=refresh]` fail (not likely, but certainly not impossible.) Should it go before or after the stream error redirect markup? My vote is for before.

Reply:
@tigt I might've missed this in another part of this discussion, but is there value in writing both the meta and script tags here? In which case would one work but not the other? I feel like meta is more practically reliable and works with js disabled which is a bonus.

Reply:
I'll also add that having dangling markup is probably very rare since for the most part things are escaped. To have dangling markup it would mean that you would have gotten your application into a state where you are passing invalid markup to an unescaped interpolation which is an issue in and of itself.

Reply:
I don’t know of a specific browser where it happens, but my intuition makes me worry about relying solely on the option that is technically invalid HTML — all it would take is one high-profile unescaped markup attack to make some security-conscious browser ignore meta refreshes in the `<body>`. (Maybe Tor Browser or Brave already do.)

Reply:
Blast from the future: turns out there _might_ be value in writing both `meta[http-equiv="refresh"]` and `window.location=`. [WebKit changeset 280,870](https://trac.webkit.org/changeset/280870/webkit/):

>  Meta HTTP refresh should not navigate if document has sandboxed automatic features browsing context flag set:
>
> * ​https://html.spec.whatwg.org/multipage/semantics.html#shared-declarative-refresh-steps (Step 13) 
>
> Firefox and Chrome already behave this way.

I haven’t managed to figure out what the spec _means_ by that yet, but putting it here for later.